packedForest/src/forestTypes/unsupervisedForests/urf/urfTree.h

```diff
@@ -0,0 +1,267 @@
+#ifndef urfTree_h
+#define urfTree_h
+#include "../../../baseFunctions/fpBaseNode.h"
+#include "unprocessedURFNode.h"
+#include <vector>
+#include <map>
+#include <assert.h>
+#include <random>
+#include <eigen3/Eigen/Dense>
+#include <eigen3/Eigen/Sparse>
+#include <eigen3/Eigen/Core>
+
+typedef Eigen::SparseMatrix<double> SpMat;
+typedef Eigen::Triplet<double> TripType;
+
+namespace fp{
+
+	template <typename T>
+		class urfTree
+		{
+			protected:
+				float OOBAccuracy;
+				float correctOOB;
+				float totalOOB;
+				std::vector< fpBaseNode<T, int> > tree;
+				std::vector< unprocessedURFNode<T> > nodeQueue;
+				std::vector< unprocessedURFNode<T> > leafNodes;
+
+			public:
+				urfTree() : OOBAccuracy(-1.0),correctOOB(0),totalOOB(0){}
+
+				void loadFirstNode(){
+					nodeQueue.emplace_back(fpSingleton::getSingleton().returnNumObservations());
+				}
+
+				inline bool shouldProcessNode(){
+					if(nodeQueue.back().returnNodeImpurity()<=0){
+						return false;
+					}
+					if(nodeQueue.back().returnInSampleSize() <= fpSingleton::getSingleton().returnMinParent()){
+						
+return false;
+					}
+                if(nodeQueue.back().returnDepth() >= fpSingleton::getSingleton().returnMaxDepth()){
+                          return false;
+                      }
+					return true;
+				}
+
+				inline float returnOOB(){
+					return correctOOB/totalOOB;
+				}
+
+				inline int returnLastNodeID(){
+					return tree.size()-1;
+				}
+
+				inline void linkParentToChild(){
+					if(nodeQueue.back().returnIsLeftNode()){
+						tree[nodeQueue.back().returnParentID()].setLeftValue(returnLastNodeID());
+					}else{
+						tree[nodeQueue.back().returnParentID()].setRightValue(returnLastNodeID());
+					}
+				}
+
+
+				inline int returnMaxDepth(){
+					int maxDepth=0;
+					for(auto nodes : tree){
```

Comment:
side effect perhaps of using rerftree as base, but i noticed that each of these probably should be using `auto &`

Reply:
sure